### PR TITLE
SQL: enhance IN operator support for formatted date fields. Extend allowed conversions

### LINKED ISSUE #58932: SQL: IN operator incorrectly handling specified date format
Given a field with the mapping:
```
"datetime0" : {
    "type" : "date",
    "format" : "yyyy-MM-dd' 'HH:mm:ss"
}
```
and a query with the `IN` filtering: `SELECT datetime0 FROM idx WHERE datetime0 IN ('2000-11-11 11:11:11'::timestamp)` will raise an exception:
```
"root_cause" : [
      {
        "type" : "parse_exception",
        "reason" : "failed to parse date field [2000-11-11T11:11:11Z] with format [yyyy-MM-dd' 'HH:mm:ss]: [failed to parse date field [2000-11-11T11:11:11Z] with format [yyyy-MM-dd' 'HH:mm:ss]]"
      }
    ]
```
The query executes correctly with the equality operator.
Data in the index is irrelevant (the index can contain no documents).

### PR DESCRIPTION
**Pull request by bpintea**
The `IN` operator fails the comparison against `datetime` values, if these have a custom format; reason being that translation of these values is unaware of the field mapping format and just use the `strict_date_time` one in a `terms` query.
To overcome this limitation the PR changes the translation to a boolean `should` of `range`s query, which allow providing a `format` parameter (which still won't be that of the mapped field, but one that allows ES to interpret the provided values of the range).

The PR also adjusts the way `IN` converts the values in the set, in case the clause uses a field attribute: currently, the type of the first value in the `IN` set is considered as the reference and all subsequent values converted to it. This is changed to take the type of the field attribute instead as the reference. 
The data type validation between the operands of the `IN` is also changed from a compatibility test to a convertibility test (which is what the query generation eventually does anyways).
These two changes allow for a wide range of implicit conversions (like `WHERE datetime_field IN ('datetime_literal'::datetime, 'date_literal'::date, 'datetime_literal')`, what other implementation also support).

Closes #58932.

Co-authored-by: yinanwu yinanwu@tencent.com

### CODE REVIEW

**Comment by bpintea** [comment]
@elasticmachine update branch

File: x-pack/plugin/sql/qa/server/src/main/resources/filter.csv-spec [hunk 1]
```diff
@@ -131,21 +131,22 @@ SELECT COUNT(*), TRUNCATE(emp_no, -2) t FROM test_emp WHERE 'aaabbb' RLIKE 'a{2,
 ;
 
 inWithCompatibleDateTypes
-SELECT birth_date FROM test_emp WHERE birth_date IN ({d '1959-07-23'},CAST('1959-12-25T12:12:12' AS TIMESTAMP)) OR birth_date IS NULL ORDER BY birth_date;
```
Comment (bpintea) [inline x-pack/plugin/sql/qa/server/src/main/resources/filter.csv-spec:134]: This test - and logic allowing it - was faulty.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java [hunk 2]
```diff
@@ -665,14 +665,10 @@ public void testHavingOnScalar() {
                 error("SELECT int FROM test GROUP BY int HAVING 2 < ABS(int)"));
     }
 
-    public void testInWithDifferentDataTypes() {
-        assertEquals("1:8: 2nd argument of [1 IN (2, '3', 4)] must be [integer], found value ['3'] type [keyword]",
```
Comment (bpintea) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java:669]: This test is now possible (and inline with what other SQL implementations do).

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java [hunk 3]
```diff
@@ -391,18 +391,32 @@ public static Query doTranslate(In in, TranslatorHandler handler) {
...
+                });
+
+                if (dt == DATETIME) {
+                    DateFormatter formatter = DateFormatter.forPattern(DATE_FORMAT);
+
+                    q = null;
+                    for (Object o : set) {
+                        assert o instanceof ZonedDateTime;
+                        // see comment in Ranges#doTranslate() as to why formatting as String is required
+                        String zdt = formatter.format((ZonedDateTime) o);
```
Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java:412]: @matriv, related to your [suggestion](https://github.com/elastic/elasticsearch/pull/60377#issuecomment-681878439), there are similarities with `BinaryComparisons`'s times values code, but not sure if enough for a common extraction to make sense (despite the comment reference). But happy to further iterate.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java [hunk 4]
```diff
@@ -391,18 +391,32 @@ public static Query doTranslate(In in, TranslatorHandler handler) {
...
+                        set.add(handler.convert(valueOf(e), dt));
+                    }
+                });
+
+                if (dt == DATETIME) {
+                    DateFormatter formatter = DateFormatter.forPattern(DATE_FORMAT);
+
+                    q = null;
+                    for (Object o : set) {
+                        assert o instanceof ZonedDateTime;
```
Comment (matriv) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java:410]: minor, but better add a short msg.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java (hunk 3 above)
Comment (matriv) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java:412]: Yeah, seems find as is.

**Review by matriv — COMMENTED** [review]
Everything seems good, thx!
Could you please check and maybe remove this part: https://github.com/elastic/elasticsearch/pull/63483/files#diff-913993cd17dbceac1fa55d21b31c61deR291

**Comment by bpintea** [comment]
> Could you please check and maybe remove this part: https://github.com/elastic/elasticsearch/pull/63483/files#diff-913993cd17dbceac1fa55d21b31c61deR291

Not sure if that was the intended link?
If you meant the previously mentioned [else-branch](https://github.com/elastic/elasticsearch/blob/2b94f88639dbf142576a1f24af34156b033190b2/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java#L246), I had already checked it and it is indeed reachable (ex.: `.. time_field::time = 'literal'::time`), so all good there.

**Comment by matriv** [comment]
> > Could you please check and maybe remove this part: https://github.com/elastic/elasticsearch/pull/63483/files#diff-913993cd17dbceac1fa55d21b31c61deR291
> 
> Not sure if that was the intended link?
> If you meant the previously mentioned [else-branch](https://github.com/elastic/elasticsearch/blob/2b94f88639dbf142576a1f24af34156b033190b2/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java#L246), I had already checked it and it is indeed reachable (ex.: `.. time_field::time = 'literal'::time`), so all good there.

Indeed that was the link THX!, do we have some tests already for `field::time IN ('.....'::time, ...)` ?

**Comment by bpintea** [comment]
@elasticmachine update branch

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java (hunk 4 above)
Comment (bpintea) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java]: Sure, added.

**Comment by bpintea** [comment]
> do we have some tests already for `field::time IN ('.....'::time, ...)` ?

No. 
It's a good idea to add it, but that currently that fails because of #63521. But that should be fix probably in a subsequent PR, the construct above doesn't access the code this PR toches.

**Review by matriv — APPROVED** [review]
LGTM, Thx for checking @bpintea!

**Review by astefan — APPROVED** [review]
LGTM

**Comment by bpintea** [comment]
> do we have some tests already for `field::time IN ('.....'::time, ...)` ?

I've closed #63521, this test can now run, so I've added one.

**Comment by palesz** [comment]
@bpintea Since this also fixes #64238, I think we should backport this to 7.10 (or even 7.9).

**Comment by bpintea** [comment]
> @bpintea Since this also fixes #64238, I think we should backport this to 7.10 (or even 7.9).

It does, but it also enlarges a bit the allowed functionality due to the allowed implicit conversions (besides now working with formatted dates), so it's arguably more of a feature coming with an inherent fix, strictly speaking. 
Still, I'd agree that it'd be worth the back port to 7.10 (7.9 branch won't have any new releases), but not my call. @costin?

**Pull request merged**